### ISSUE TITLE
Porcelain pull force mode triggered KeyError

### ISSUE THREAD

**Issue by Kiramei** [jelmer/dulwich#1399]
I use the dulwich in my project and originally I mean to keep the local repository synchronized with the remote repository. Since I tried to get the information mentioned in the previous issues, the problem of `dulwich.porcelain.DivergedBranches` which indicates that the local sha doesn't match with the remote seems to be concerning an unaddressed problem that the merge of the pull can't be solved until the moment. Well, after that, I turned to the command corresponding `git pull --force` like:
```python
  porcelain.pull(repo, REPO_URL_HTTP, 'master', force=True)
```
to settle this problem, but it turned out to be a case of KeyError.

This is the message of the traceback :
```PlainText
Traceback (most recent call last):
  File "c:\Program Files\xxxxxxxxxxxx\test_in.py", line 342, in check_install
    check_git(opt)
  File "c:\Program Files\xxxxxxxxxxxx\test_in.py", line 314, in check_git
    porcelain.pull(repo, REPO_URL_HTTP, 'master', force=True, no_ff=True)
  File "C:\Users\xxxxx\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\porcelain.py", line 1349, in pull
    tree = r[b"HEAD"].tree
  File "C:\Users\xxxxx\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\repo.py", line 839, in __getitem__
    return self.object_store[self.refs[name]]
  File "C:\Users\xxxxx\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\object_store.py", line 150, in __getitem__
    type_num, uncomp = self.get_raw(sha1)
  File "C:\Users\xxxxx\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\object_store.py", line 597, in get_raw
    raise KeyError(hexsha)
KeyError: b'aa35e3e861afd7c26171c9e1b3530647a930b0d8'
```

In my case, the work can be concluded as: The remote repository keeps updating, and the local repository should be updated with the remote, while the local changes whether stashed or unstashed are to be dropped. I'd appreciate it if there's someone that can give me a suggestion on it.

**Comment by jelmer** [comment]
Have you done any debugging - is aa35e3e861afd7c26171c9e1b3530647a930b0d8 the correct commit sha? Did it get fetched to the local repository?

**Comment by Kiramei** [comment]
`aa35e3e861afd7c26171c9e1b3530647a930b0d8` is the remote latest SHA. If the `fetch ` process is done, the local sha will be forced to change to the remote key, but the local repository files doesn't change to get synchronized with the remote one.

**Comment by Kiramei** [comment]
In my case, I also find that the process is all fine if I use Dulwich to clone the project at the beginning, while my problem occurs when I try moving the script into the local repository which was cloned by `Github Cli` and outdated.

**Comment by jelmer** [comment]
If you run porcelain pull, does the commit actually get fetched?

**Comment by Kiramei** [comment]
It seems there're some problems in my pull process and it seems not properly fetched. 

Here's my debug prompt:

```PlainText
Python 3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.

>>> from dulwich import porcelain
>>> repo = Repo('.')
>>> porcelain.status(repo)

GitStatus(staged={'add': [], 'delete': [], 'modify': []}, unstaged=[b'README.md', b'device_operation/auto_scan_simulator.py', b'main.py', b'module/cafe_reward.py', b'requirements.txt'], untracked=[])

>>> REPO_URL_HTTP = 'https://gitee.com/pur1fy/blue_archive_auto_script.git'
>>> porcelain.pull(repo, REPO_URL_HTTP, 'master')

Traceback (most recent call last):
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\porcelain.py", line 1338, in pull
    check_diverged(r, r.refs.follow(rh)[1], fetch_result.refs[lh])
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\porcelain.py", line 341, in check_diverged
    raise DivergedBranches(current_sha, new_sha)
dulwich.porcelain.DivergedBranches: (b'72116c892d67b4904e54a0bc04467161d2714a62', b'812c466da8ed18a45327ceba58e7cf0e3f932d75')

>>> porcelain.fetch(repo, REPO_URL_HTTP)

FetchPackResult({b'HEAD': b'812c466da8ed18a45327ceba58e7cf0e3f932d75', b'refs/heads/Uooc-script': b'30f4168d82b16607c7c2a5b7d3364df847febf72', b'refs/heads/fix_device_operation': b'eb03dc756f4c30ed681198e9e4a89c95494048f1', b'refs/heads/gh-pages': b'954fabffdeac5918f853b82fdfbcdf8f2045dd65', b'refs/heads/master': b'812c466da8ed18a45327ceba58e7cf0e3f932d75', b'refs/heads/pur1fy_dev': b'a5ea5f0d870187b473dc7f1b08227de4e369c9de', b'refs/heads/ui_dev': b'bcc34e11013f8590b012536f9d5e8a4f07d289c9', b'refs/tags/Update': b'f64924e13ab8f3da1684620c7b78bc4e6fb21d0d', b'refs/tags/beta': b'8d28e03f217c1b1b3a5aebd5658e69117604d849', b'refs/tags/v1.1.0': b'60d0114e94b31d1dd7dfa2ab1362daf86ff1059c', b'refs/tags/v1.1.1': b'18595189a99950e2739ab59d9cd930a3a40d5bbd', b'refs/tags/v1.1.2': b'8447b44d65d89250bb63d0b58748ccf180dcc046', b'refs/tags/v1.1.3': b'18595189a99950e2739ab59d9cd930a3a40d5bbd'}, {b'HEAD': b'refs/heads/master'}, b'git/2.43.0\n')

>>> porcelain.pull(repo, REPO_URL_HTTP, 'master')

  File "<stdin>", line 1, in <module>
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\porcelain.py", line 1338, in pull
    check_diverged(r, r.refs.follow(rh)[1], fetch_result.refs[lh])
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\porcelain.py", line 341, in check_diverged
    raise DivergedBranches(current_sha, new_sha)
dulwich.porcelain.DivergedBranches: (b'72116c892d67b4904e54a0bc04467161d2714a62', b'812c466da8ed18a45327ceba58e7cf0e3f932d75')
>>> porcelain.pull(repo, REPO_URL_HTTP, 'master', force=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\porcelain.py", line 1349, in pull
    tree = r[b"HEAD"].tree
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\repo.py", line 839, in __getitem__
    return self.object_store[self.refs[name]]
    type_num, uncomp = self.get_raw(sha1)
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\object_store.py", line 597, in get_raw
    raise KeyError(hexsha)
KeyError: b'812c466da8ed18a45327ceba58e7cf0e3f932d75'

>>> porcelain.reset(repo, mode='hard')

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\porcelain.py", line 1164, in reset
    tree = parse_tree(r, treeish)
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\objectspec.py", line 66, in parse_tree
    o = repo[treeish]
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\repo.py", line 839, in __getitem__
    return self.object_store[self.refs[name]]
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\object_store.py", line 150, in __getitem__
    type_num, uncomp = self.get_raw(sha1)
  File "C:\Program Files\BlueArchiveAutoScript\env\lib\site-packages\dulwich\object_store.py", line 597, in get_raw
    raise KeyError(hexsha)
KeyError: b'812c466da8ed18a45327ceba58e7cf0e3f932d75'
```

**Comment by jelmer** [comment]
What version of Dulwich are you using? Can you try running this code with protocol_version=0 as argument to porcelain.pull?

**Comment by Kiramei** [comment]
## Thank you very much !!
My Dulwich version is: 0.22.3, and I fixed my problem with your suggestion "running this code with protocol_version=0 as argument to `porcelain.pull` ". It now works well, and I wonder why the small fix can settle the problem down.

``` python
>>> porcelain.pull(repo, REPO_URL_HTTP, 'master', protocol_version=0)

Enumerating objects: 895, done.
Counting objects: 100% (822/822), done.
Compressing objects: 100% (415/415), done.
Total 687 (delta 449), reused 505 (delta 270), pack-reused 0
copied 686 pack entries86/687
```

**Comment by jelmer** [comment]
Cc @stspdotname 

Since we have two open bugs about V2 support I wonder if it would be prudent to just default to v0 for now and make v2 and opt-in feature until these are resolved.

**Comment by stspdotname** [comment]
Sure. We're not going to be finding issues like this if nobody tests v2 though. Making it opt-in might discourage such testing.

I will take a closer look at this problem soon.

**Comment by stspdotname** [comment]
I have tested the above reproduction recipe with dulwich code from the current master branch, on a an artificially backdated clone of the repository with a dummy commit added on the master branch to forge diverged history.

It looks like the fixes for issue #1389 also helped here, see trace below.
@Kiramei , can you confirm that the problem also appears fixed at your end?

```
>>> porcelain.fetch(repo, url)
FetchPackResult({b'HEAD': b'9e4dd42e2a6483a3547b5e0d031fa98cbdabc7ae', b'refs/heads/Uooc-script': b'30f4168d82b16607c7c2a5b7d3364df847febf72', b'refs/heads/fix_device_operation': b'eb03dc756f4c30ed681198e9e4a89c95494048f1', b'refs/heads/gh-pages': b'556584fb81016128ee9f7270b43baf1875deae7e', b'refs/heads/master': b'9e4dd42e2a6483a3547b5e0d031fa98cbdabc7ae', b'refs/heads/pur1fy_dev': b'a5ea5f0d870187b473dc7f1b08227de4e369c9de', b'refs/heads/ui_dev': b'958c6f7aef8b7af588b1f4c1b83249ad4e527611', b'refs/tags/Update': b'f64924e13ab8f3da1684620c7b78bc4e6fb21d0d', b'refs/tags/beta': b'8d28e03f217c1b1b3a5aebd5658e69117604d849', b'refs/tags/v1.1.0': b'60d0114e94b31d1dd7dfa2ab1362daf86ff1059c', b'refs/tags/v1.1.1': b'18595189a99950e2739ab59d9cd930a3a40d5bbd', b'refs/tags/v1.1.2': b'8447b44d65d89250bb63d0b58748ccf180dcc046', b'refs/tags/v1.1.3': b'18595189a99950e2739ab59d9cd930a3a40d5bbd'}, {b'HEAD': b'refs/heads/master'}, b'git/2.43.0\n')
>>> porcelain.pull(repo, url, 'master')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/stsp/.virtualenvs/dulwich/lib/python3.11/site-packages/dulwich/porcelain.py", line 1338, in pull
    check_diverged(r, r.refs.follow(rh)[1], fetch_result.refs[lh])
  File "/home/stsp/.virtualenvs/dulwich/lib/python3.11/site-packages/dulwich/porcelain.py", line 341, in check_diverged
    raise DivergedBranches(current_sha, new_sha)
dulwich.porcelain.DivergedBranches: (b'e0558d7d810d1897247c24f0d3fd856ab895e3b4', b'9e4dd42e2a6483a3547b5e0d031fa98cbdabc7ae')
>>> porcelain.pull(repo, url, 'master', force=True)
>>> porcelain.reset(repo, mode='hard')
>>> 
```

**Comment by Kiramei** [comment]
@stspdotname  I've tried your recipe, and I was sorry that I can't get the same result as yours, which could also be a problem of V2 error.

This is the trace:

```
>>> porcelain.fetch(repo, url)
FetchPackResult({b'HEAD': b'dfa2f171c8b46c586ac22837477ac5905ab41316', b'refs/heads/Uooc-script': b'30f4168d82b16607c7c2a5b7d3364df847febf72', b'refs/heads/fix_device_operation': b'eb03dc756f4c30ed681198e9e4a89c95494048f1', b'refs/heads/gh-pages': b'cd30facee016e883b00ecf3c5eecf05e196b84e2', b'refs/heads/master': b'dfa2f171c8b46c586ac22837477ac5905ab41316', b'refs/heads/pur1fy_dev': b'a5ea27611', b'refs/tags/Update': b'f64924e13ab8f3da1684620c7b78bc4e6fb21d0d', b'refs/tags/beta': b'8d28e03f217c1b1b3a5aebd5658e69117604d849', b'refs/tags/v1.1.0': b'60d0114e94b31d1dd7dfa2ab1362daf86ff1059c', b'refs/tags/v1.1.1': b'18595189a99950e2739ab59d9cd930a3a40d5bbd', b'refs/tags/v1.1.2': b'8447b44d65d89250bb63d0b58748ccf180dcc046', b'refs/tags/v1.1.3': b'18595189a99950e2739ab59d9cd930a3a40d5bbd'}, {b'HEAD': b'refs/heads/master'}, b'git/2.43.0\n')
>>> porcelain.pull(repo, url, 'master')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\porcelain.py", line 1338, in pull
    check_diverged(r, r.refs.follow(rh)[1], fetch_result.refs[lh])
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\porcelain.py", line 341, in check_diverged
    raise DivergedBranches(current_sha, new_sha)
dulwich.porcelain.DivergedBranches: (b'72116c892d67b4904e54a0bc04467161d2714a62', b'dfa2f171c8b46c>>> porcelain.pull(repo, url, 'master', force=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\porcelain.py", line 1349, in pull
    tree = r[b"HEAD"].tree
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\repo.py", line 839, in __getitem__
    return self.object_store[self.refs[name]]
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\object_store.py", line 150, in __getitem__
    type_num, uncomp = self.get_raw(sha1)
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\object_store.py", line 597, in get_raw
    raise KeyError(hexsha)
KeyError: b'dfa2f171c8b46c586ac22837477ac5905ab41316'
>>> porcelain.reset(repo, mode='hard')
Traceback (most recent call last):
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\porcelain.py", line 1164, in reset
    tree = parse_tree(r, treeish)
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\objectspec.py", line 66, in parse_tree
    o = repo[treeish]
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\repo.py", line 839, in __getitem__
    return self.object_store[self.refs[name]]
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\object_store.py", line 150, in __getitem__
    type_num, uncomp = self.get_raw(sha1)
  File "C:\Users\m1710\AppData\Local\Programs\Python\Python310\lib\site-packages\dulwich\object_store.py", line 597, in get_raw
    raise KeyError(hexsha)
KeyError: b'dfa2f171c8b46c586ac22837477ac5905ab41316'
>>>
```

**Comment by stspdotname** [comment]
Are you trying this on a fresh clone of the repository?

Object ID dfa2f171c8b46c586ac22837477ac5905ab41316 does not exist in the repository I cloned from gitee.com.
Does this object actually exist in your local repository?

I am wondering whether you are now seeing some side-effect of the failed force-push operation run with the buggy version of Dulwich?

**Comment by Kiramei** [comment]
dfa2f171c8b46c586ac22837477ac5905ab41316 is the relatively latest version (11/1 18:00) of the repo, and I feel sorry that I missed a detail for you that our repo is developed on GitHub but for the speed of version check, we make a mirror on Gitee. Though it might take some time for Gitee to fetch the latest version, it still works well with V1. Certainly, our .git directory is always corresponding to Gitee one, and you're right that your recipe works well for a fresh clone. For older version to update, our previous method was using Git Cli and everything is ok.

**Comment by stspdotname** [comment]
To make progress we need to understand whether the KeyError you are seeing is due to a problem in your local repository, or whether there is still a problem in Dulwich.

Everything seems to be working as expected on a fresh clone, correct? Or can you show a reproduction recipe which triggers the KeyError on a fresh clone of the repository? If the problem cannot be reproduced on a fresh clone then we won't be able to analyze the issue and provide a fix. Unless we can somehow get access to a copy of your specific clone where the problem can be triggered (in a zip archive or similar).

**Comment by Kiramei** [comment]
Thanks for your comment. Here I create a package download URL for you to debug: https://github.com/Kiramei/Kiramei/releases/download/Debug/blue_archive_auto_script-72116c892d67b4904e54a0bc04467161d2714a62.tar.gz . You may have the recipe you've tried before in the repo and try debugging it.

**Comment by stspdotname** [comment]
> Thanks for your comment. Here I create a package download URL for you to debug: https://github.com/Kiramei/Kiramei/releases/download/Debug/blue_archive_auto_script-72116c892d67b4904e54a0bc04467161d2714a62.tar.gz . You may have the recipe you've tried before in the repo and try debugging it.

The reference refs/remotes/origin/master points at commit 8447b44d65d89250bb63d0b58748ccf180dcc046.

This commit has parent commit 38c08e5506857ba6bf978c74b76f4afe283e7850 and this commit can be found nowhere, neither as a loose object nor in any of the pack files inside the repository.

git log reveals that this repository is a shallow clone (the origin/master tip commit is grafted):
```
$ git log refs/remotes/origin/master
commit 8447b44d65d89250bb63d0b58748ccf180dcc046 (grafted, origin/master, origin/HEAD)
```
When I ask Dulwich to force-pull with Git protocol version 0 then Dulwich receives a pack from the server which contains the missing objects. So it seems we still have problems with shallow repositories in combination with protocol v2.

**Comment by jelmer** [comment]
@stspdotname Ah, thanks for debugging that. Any chance you could file a separate bug about that - are you working on it?

**Comment by jelmer** [comment]
That test is incorrect - you can't use depth without setting up shallows.

**Comment by stspdotname** [comment]
I have tweaked my test case, which is now passing across all protocols and protocol versions, see above commit.
This test case shows that deepening a shallow repository is possible and working, but we need to request an infinite depth to avoid the KeyError. Otherwise, the clone remains shallow and it is expected that certain objects remain missing and yield KeyErrors when accessed.

@Kiramei could you please check whether adding depth=dulwich.protocol.DEPTH_INFINITE to your call to porcelain.fetch() makes it work?

**Comment by Kiramei** [comment]
@stspdotname It  works well.

```plaintext
>>> import dulwich
>>> from dulwich import porcelain
>>> from dulwich.repo import Repo
>>> repo = Repo('.')
>>> porcelain.fetch(repo, "https://gitee.com/pur1fy/blue_archive_auto_script.git", depth=dulwich.protocol.DEPTH_INFINITE)
Enumerating objects: 30941, done.
copied 30940 pack entries40/30941
FetchPackResult({b'HEAD': b'5f205783b9899c3a91b94a4422d3be1164bd44f0', b'refs/heads/Daodanfd5-patch-1': b'5f6c173c00b3ec6bdc4d593ba34e0a77a62a3e79', b'refs/heads/Uooc-script': b'30f4168d82b16607c7c2a5b7d3364df847febf72', b'refs/heads/create-ui-fix': b'eab7a8d1ae9f35bf76f796365618320bc67bdb26', b'refs/heads/gh-pages': b'a1ec8a5313732ce487ff5ce85f3a73adc12febfa', b'refs/heads/master': b'5f205783b9899c3a91b94a4422d3be1164bd44f0', b'refs/heads/pur1fy_dev': b'e4528295d054c9c9186143d65d4cb2071cec8bbd', b'refs/heads/ui_dev': b'dde011500264933dcfcc96ad1799b07d14c59847', b'refs/tags/Update': b'f64924e13ab8f3da1684620c7b78bc4e6fb21d0d', b'refs/tags/beta': b'8d28e03f217c1b1b3a5aebd5658e69117604d849', b'refs/tags/v1.1.0': b'60d0114e94b31d1dd7dfa2ab1362daf86ff1059c', b'refs/tags/v1.1.1': b'18595189a99950e2739ab59d9cd930a3a40d5bbd', b'refs/tags/v1.1.2': b'8447b44d65d89250bb63d0b58748ccf180dcc046', b'refs/tags/v1.1.3': b'18595189a99950e2739ab59d9cd930a3a40d5bbd'}, {b'HEAD': b'refs/heads/master'}, b'git/2.43.0\n')
>>> porcelain.pull(repo,  "https://gitee.com/pur1fy/blue_archive_auto_script.git", 'master')
>>> porcelain.reset(repo, mode='hard')
```

**Comment by stspdotname** [comment]
Glad to hear this, thank you for checking!
